xde/src/mac.rs

```diff
@@ -110,30 +168,30 @@ impl MacClient {
         unsafe { mac_rx_set(self.mch, rx_fn, arg) };
     }
 
+    /// Register promiscuous callback to receive packets on the underlying MAC.
     pub fn add_promisc(
-        &self,
+        self: &Arc<Self>,
         ptype: mac_client_promisc_type_t,
         promisc_fn: mac_rx_fn,
-        arg: *mut c_void,
         flags: u16,
-    ) -> Result<*mut mac_promisc_handle, c_int> {
+    ) -> Result<MacPromiscHandle, c_int> {
         let mut mph = 0 as *mut mac_promisc_handle;
 
+        // This extra ref is cleaned up in `MacPromiscHandle::drop()`
+        let arg = Arc::into_raw(self.clone()) as *mut c_void;
```

Comment:
Can we avoid the extra ref by just doing `let arg = Arc::as_ptr(self) as *mut c_void` here?